File: backend/src/main/java/com/raster/hrm/shift/service/ShiftService.java
package com.raster.hrm.shift.service;

import com.raster.hrm.exception.BadRequestException;
import com.raster.hrm.exception.ResourceNotFoundException;
import com.raster.hrm.shift.dto.ShiftRequest;
import com.raster.hrm.shift.dto.ShiftResponse;
import com.raster.hrm.shift.entity.Shift;
import com.raster.hrm.shift.entity.ShiftType;
import com.raster.hrm.shift.repository.ShiftRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Transactional
public class ShiftService {

    private static final Logger log = LoggerFactory.getLogger(ShiftService.class);

    private final ShiftRepository shiftRepository;

    public ShiftService(ShiftRepository shiftRepository) {
        this.shiftRepository = shiftRepository;
    }

    @Transactional(readOnly = true)
    public Page<ShiftResponse> getAll(Pageable pageable) {
        return shiftRepository.findAll(pageable)
                .map(this::mapToResponse);
    }

    @Transactional(readOnly = true)
    public ShiftResponse getById(Long id) {
        var shift = shiftRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Shift", "id", id));
        return mapToResponse(shift);
    }

    @Transactional(readOnly = true)
    public List<ShiftResponse> getByType(ShiftType type) {
        return shiftRepository.findByType(type).stream()
                .map(this::mapToResponse)
                .toList();
    }

    @Transactional(readOnly = true)
    public List<ShiftResponse> getActive() {
        return shiftRepository.findByActive(true).stream()
                .map(this::mapToResponse)
                .toList();
    }

    public ShiftResponse create(ShiftRequest request) {
        if (shiftRepository.existsByName(request.name())) {
            throw new BadRequestException("Shift with name '" + request.name() + "' already exists");
        }

        var shift = new Shift();
        shift.setName(request.name());
        shift.setType(ShiftType.valueOf(request.type()));
        shift.setStartTime(request.startTime());
        shift.setEndTime(request.endTime());
        shift.setBreakDurationMinutes(request.breakDurationMinutes() != null ? request.breakDurationMinutes() : 0);
        shift.setGracePeriodMinutes(request.gracePeriodMinutes() != null ? request.gracePeriodMinutes() : 0);
        shift.setDescription(request.description());

        var saved = shiftRepository.save(shift);
        log.info("Created shift with id: {} name: {}", saved.getId(), saved.getName());
        return mapToResponse(saved);
    }

    public ShiftResponse update(Long id, ShiftRequest request) {
        var shift = shiftRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Shift", "id", id));

        if (!shift.getName().equals(request.name()) && shiftRepository.existsByName(request.name())) {
            throw new BadRequestException("Shift with name '" + request.name() + "' already exists");
        }

        shift.setName(request.name());
        shift.setType(ShiftType.valueOf(request.type()));
        shift.setStartTime(request.startTime());
        shift.setEndTime(request.endTime());
        shift.setBreakDurationMinutes(request.breakDurationMinutes() != null ? request.breakDurationMinutes() : 0);
        shift.setGracePeriodMinutes(request.gracePeriodMinutes() != null ? request.gracePeriodMinutes() : 0);
        shift.setDescription(request.description());

        var saved = shiftRepository.save(shift);
        log.info("Updated shift with id: {}", saved.getId());
        return mapToResponse(saved);
    }

    public ShiftResponse updateActive(Long id, boolean active) {
        var shift = shiftRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Shift", "id", id));

        shift.setActive(active);
        var saved = shiftRepository.save(shift);
        log.info("Updated active status of shift id: {} to {}", id, active);
        return mapToResponse(saved);
    }

    public void delete(Long id) {
        var shift = shiftRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Shift", "id", id));
        shiftRepository.delete(shift);
        log.info("Deleted shift with id: {}", id);
    }

    private ShiftResponse mapToResponse(Shift shift) {
        return new ShiftResponse(
                shift.getId(),
                shift.getName(),
                shift.getType().name(),
                shift.getStartTime(),
                shift.getEndTime(),
                shift.getBreakDurationMinutes(),
                shift.getGracePeriodMinutes(),
                shift.getDescription(),
                shift.isActive(),
                shift.getCreatedAt(),
                shift.getUpdatedAt()
        );
    }
}

File: frontend/src/app/services/shift.service.ts
import { Injectable } from '@angular/core';
import { HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ApiService } from './api.service';
import { ShiftRequest, ShiftResponse } from '../models/shift.model';
import { Page } from '../models/page.model';

@Injectable({ providedIn: 'root' })
export class ShiftService {
  private path = '/shifts';

  constructor(private api: ApiService) {}

  getAll(page = 0, size = 10): Observable<Page<ShiftResponse>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.api.get<Page<ShiftResponse>>(this.path, params);
  }

  getById(id: number): Observable<ShiftResponse> {
    return this.api.get<ShiftResponse>(`${this.path}/${id}`);
  }

  getByType(type: string): Observable<ShiftResponse[]> {
    return this.api.get<ShiftResponse[]>(`${this.path}/type/${type}`);
  }

  getActive(): Observable<ShiftResponse[]> {
    return this.api.get<ShiftResponse[]>(`${this.path}/active`);
  }

  create(shift: ShiftRequest): Observable<ShiftResponse> {
    return this.api.post<ShiftResponse>(this.path, shift);
  }

  update(id: number, shift: ShiftRequest): Observable<ShiftResponse> {
    return this.api.put<ShiftResponse>(`${this.path}/${id}`, shift);
  }

  updateActive(id: number, active: boolean): Observable<ShiftResponse> {
    return this.api.patch<ShiftResponse>(`${this.path}/${id}/active`, { active });
  }

  delete(id: number): Observable<void> {
    return this.api.delete<void>(`${this.path}/${id}`);
  }
}
